### PR TITLE
win32: su -c CMD: improvements

### PR DESCRIPTION
**Pull request by avih**
First commit fixes an issue which was first mentioned [here](https://github.com/rmyorston/busybox-w32/issues/314#issuecomment-1514491470), where CMD was not constructed correctly for `ShellExecute`, as it was lacking escape for double-quotes and (some) backslashes, which could result in incorrect value and/or multiple parsed args instead of one.

The second commit allows adding arguments as `su -c CMD [ARG...]` so that it's easier to use values from the current shell instead of quoting them for shell-re-input and embedding them in CMD.

### CODE REVIEW

**Comment by rmyorston** [comment]
Thanks for persevering with this.

I've put a modified version on my su_cmd branch.

There's already a function to quote arguments, `quote_arg()` in win32/process.h, so I used that.

I also exported a function to append a string to a list from my `make` implementation.

**Comment by rmyorston** [comment]
>Does 64 sound reasonable to you?

It seems reasonable. The first patch added 0 bytes. Exporting `xappendword()` added 32 for the 32-bit build, subtracted 16 for the 64-bit build. The results for the third patch were:
```
function                                             old     new   delta
suw32_main                                           272     336     +64
packed_usage                                       13744   13760     +16
.rdata                                             53360   53344     -16
------------------------------------------------------------------------------
(add/remove: 0/0 grow/shrink: 2/1 up/down: 80/-16)             Total: 64 bytes
```
The usage strings are compressed.

>so consider replacing it if indeed it's equivalent?

I'll have a look. The existing function has been there since before my involvement with busybox-w32, so it must be doing something right.

The speed of `xappendword` is indeed mostly irrelevant.

**Comment by avih** [comment]
> The existing function has been there since before my involvement with busybox-w32, so it must be doing something right.

Yeah, it looks good, even if possibly slightly more code than required, and I don't recall encountring issues with it.

Anyway, overall LGTM.

**Comment by avih** [comment]
For what it's worth, building on the commits in this PR, that's a sudo-like function which forwards stdout, stderr, and the exit code back to the current shell:

```sh
sudo() {(
    [ "${1-}" ] || { >&2 echo "usage: sudo CMD [ARG...]"; exit 1; }

    set -- "$(mktemp -t bb-sudo-XXXXXX)" "$@" && [ "$1" ] || exit
    cleanup() { e=$?; rm -f -- "$1" "$1.out" "$1.err"; exit "$e"; }

    trap cleanup INT
    trap cleanup EXIT

    su -c -- \
       '(shift; "$@") >"$1.out" 2>"$1.err"; echo "$?" >"$1"; exit' \
       sudo "$@"

    while ! read e <"$1" || ! [ "$e" ]; do sleep 0.05; done
    cat <"$1.out"
    cat <"$1.err" >&2
    exit "$e"
)}
```

**Comment by ale5000-git** [comment]
At the moment I can't test it yet, but there are two test cases worth trying:
`su -c 'echo "0: $0"; echo "1: $1"; echo "2: $2"' -- my_zero '1a 1b' my_two`
`su -- sh -c 'echo "0: $0"; echo "1: $1"; echo "2: $2"' my_zero '1a 1b' my_two`

They should bothe result in:
```
0: my_zero
1: 1a 1b
2: my_two
```

**Comment by avih** [comment]
> At the moment I can't test it yet, but there are two test cases worth trying:
>    `su -c 'echo "0: $0"; echo "1: $1"; echo "2: $2"' -- my_zero '1a 1b' my_two`
>    `su -- sh -c 'echo "0: $0"; echo "1: $1"; echo "2: $2"' my_zero '1a 1b' my_two`
> 
> They should both result in:
> 
> ```
> 0: my_zero
> 1: 1a 1b
> 2: my_two
> ```

Define "should".

I think they should not, because the "template" for `su -c ` is `sh -c`, therefore I think that `su -c` should support exactly the forms which work with `sh -c`, and in your examples, if you replace `su` with `sh`, then none of them print the output you expect.

The patches here (and in rmyorston's variant) do change how `-c` is handled, to be more like `sh -c` (it's a flag without argument, but it's an error to use this flag without operands). See the second commit message.

**Comment by ale5000-git** [comment]
The su on Android (the most complete version) accept both (obviously with the addition of the user, but I haven't written it since it isn't supported on Windows).

**Comment by ale5000-git** [comment]
I just wonder, why not reuse the parameter parsing of ash?
It can be almost the same.

**Comment by avih** [comment]
> The second version (without su paramters) is also more supported.

I didn't add it because the spec was for `sh -c`, and it's the same on linux. freebsd `su` doesn't support this form either, but I don't disagree that it could be useful - which is why I wrote `sudo` above.

> I just wonder, why not reuse the parameter parsing of ash?
> It can be almost the same.

It is almost the same, but I don't see how it particularly utilizes the parameter parsing of ash.

At the `sudo` implementation above, if you remove the parts which forward back stdout/err and the exit code, and don't wait for the command to complete (which, anecdotally, is exactly what `gsudo` does, just in a much more elaborate way), then it becomes trivial (untested):
```sh
sudo() { su -c -- '"$@"' sudo "$@"; }
```
and similarly, at the `su` C code, if operands are detected without `-c`, then it could simply add the (C) arguments `"\"$@\""`, `"sudo"`, `<the rest of the operands>`. It would be very little code.

One thing worth mentioning though, is that one is then likely to expect this line to work similar to `sudo`:
```sh
su ln -s foo bar
```

However, it would not work, as `su` will complain that `-s` is unsupported.

Apparently, busybox argument parsing permutes the arguments like gnu `getopt` (which is why `ls dir/ -l` works), and is similarly affected by the override env var `POSIXLY_CORRECT`, and so will try to parse `-s` as a `su` option even when `ln` is clearly a non-option argument, which will require to modify the invocation as follows:
```sh
su -- ln -s foo bar
```

Obviously this can be solved too. It's mainly a matter of deciding how we want it to behave.

**Comment by rmyorston** [comment]
I don't think it'll ever be possible to have `sh -c` and  `su -c` work in exactly the same way.

There's a new su_cmd2 branch:
- The limited Windows version of `su` has been fixed to properly quote the command. It doesn't support specifying a user so no additional arguments are allowed.
- A new `sush` applet has been added. It behaves more like a shell, supporting both `sush -c "script" [args...]` and `sush command_file [args...]`. I'm not sure about the name: maybe it should be `raise` or `lift` to complement `drop`.

**Comment by avih** [comment]
I don't think I get the value in adding a new applet with such overlapping functionality, while still not making it fully generic for `sh` invocation.

Also, technically, I don't see how `sush FILE [ARG...]` is even allowed like the synopsis suggests. Wouldn't that cause this clause to be true `(!c_opt && command)` which will reject it and print the usage?

Also, if FILE [ARG...] works, wouldn't it also allow things like `sush -- -eux -s -c SCRIPT [ARG...]` ? because you only add `--` with c_opt, so basically the user can add arguments which `sh` can interpret as additional options.

And, at this stage, wouldn't it be simpler for `sush` to simply append all arguments, if any, to `sh -d ... -t ...` and that's it?

Then you don't need to parse options or perform any checks at all, and it's a lot simpler to understand, and you can remove `su` because such `sush` would already cover it.

The only thing you'd miss is that `-s` is not added automatically when `-c` is used (and effectively undocumented that `-s` and `-c` can be used together), and instead the user would need to add it on their own if they wish it to stay interactive. This can be added to the (inline) docs of `sh` and/or of `sush`.

**Comment by rmyorston** [comment]
>Wouldn't that cause this clause to be true (!c_opt && command) which will reject it and print the usage?

The `!c_opt && command` condition never happens because of the previous line
```
   command = c_opt ? *argv++ : NULL;
```
Happily the compiler is more observant than me and doesn't emit code for the test.

I don't want to remove the `su` applet. Though limited it matches how `su` works on Unix and is the preferred method to get an elevated shell.

The new applet provides more general functionality for people who want to write their own `sudo` script ;-)

Like `-d` and `-t` I'd prefer not to document the use of `-s` with `-c`. They're all very handy in the implementation of `su` but are non-standard.

I tried all the shells I could find on Fedora. The only one that allows `-s` and `-c` together is dash. Even upstream BusyBox ash doesn't allow it.

**Comment by avih** [comment]
> The `!c_opt && command` condition never happens because of the previous line

Right, my bad :) thanks.

> I don't want to remove the `su` applet. Though limited it matches how `su` works on Unix and is the preferred method to get an elevated shell.

But it doesn't match linux. On linux you can do this:
```sh
$ su -c 'printf "[%s]\n "$0" "$@"' root self foo bar
[self]
[foo]
[bar]
```
While the current su (also after sush was added) doesn't take arguments that I can tell, so I think at the very least this should be added to su.

> The new applet provides more general functionality

So any reason to not make it fully generic, don't parse options at all, and just append all the arguments directly to `sh -d ... -t ...` ? it would also be easier to document - just like `sh` with standard `sh` options/arguments, but elevated and without waiting for it to complete (note, it's not hard to wait for it, but there's no real value in it IMHO if it doesn't actually run/proxied into the current terminal, like `gsudo` does and to some extent (sans input) my `sudo` above).

> Like `-d` and `-t` I'd prefer not to document the use of `-s` with `-c`. They're all very handy...

What is the point of undocumented features? If you think it's handly (and I agree it is) then just let users know about it. It doesn't strictly conflict with POSIX, in the sense that POSIX does say that with `-c` the shell is non interactive, therefore in `-s` is ignored if `-s -c` is used, while not ignoring it allows adding a new functionality which was not possible before.

> The only one that allows `-s` and `-c` together is dash. Even upstream BusyBox ash doesn't allow it.

Right. For what it's worth, freebsd and netbsd `sh`, and `gwsh`  (all different derivatives of ash) also keep the shell interactive with `-s -c`.

**Comment by ale5000-git** [comment]
> I don't think it'll ever be possible to have `sh -c` and `su -c` work in exactly the same way.
> 
> There's a new su_cmd2 branch:
> 
> * The limited Windows version of `su` has been fixed to properly quote the command. It doesn't support specifying a user so no additional arguments are allowed.

Isn't possible to support specifying a user, even if only `root` is allowed?
This will make the syntax more similar to the real one: https://man7.org/linux/man-pages/man1/su.1.html

**Comment by rmyorston** [comment]
OK, there's another attempt on branch su_cmd3.

This adds support for arguments to the `-c` script and command files. To match the real `su` it requires a user name to be supplied in these cases. This must be 'root'.

**Comment by avih** [comment]
> OK, there's another attempt on branch su_cmd3.

So, if I understand correctly, the usage is one of those:
`su -c CMD_STRING [root [NAME [ARG...]]]`
`su CMD_FILE [root [ARG...]]`

If yes, then I think that the `-s -c` also needs `--` after `opt_command`, or else, (unlikely) commands which begin with `-` would be parsed incorrectly as `sh` options, which then allow the user to add arbitrary arguments to `sh`, for instance:
`su -c -eux -- root -Cf 'printf %s\\n $0 $*' name foo bar`
which would invoke, I think:
`... -s -c -eux -Cf 'printf %s\\n $0 $*' name foo bar`

And, this has unbalanced square brackets:
` "[-c 'CMD'] [root [FILE ARGS | ARG0 ARGS]"`

Other than those, LGTM.

**Comment by avih** [comment]
> Neither upstream BusyBox `su` nor `su` on Linux seem to do that. Maybe it should be considered a feature rather than a bug?

But wait, it gets worse!
A script or user which would rightly want to cover all bases for arbitrary arguments after `root`, would need to invoke it like so:
`su -c CMD_STRING  -- root -- NAME "$@"`
`su path/to/script -- root -- "$@"`

(or move the `--` to after `root`, like so: `su -c CMD_STRING root -- -- NAME "$@"`.

One `--` to stop `su` parsing option-like arguments after `root`, because it premutes by default, and one to prevent `sh` from interpreting option-like arguments as options.

**Comment by rmyorston** [comment]
Comment in upstream BusyBox `su` code:
```
    /* Note: we don't use "'+': stop at first non-option" idiom here.
     * For su, "SCRIPT ARGS" or "-c CMD ARGS" do not stop option parsing:
     * ARGS starting with dash will be treated as su options,
     * not passed to shell. (Tested on util-linux 2.28).
     */
```

**Comment by avih** [comment]
Meh. Indeed this works in `su` from util-linux
```sh
su -- root -- ./myscript -a -b
```
where one `--` goes to `su`, and the other to `sh` (and myscript sees "$@" of -a and -b).

The manpage says:
> When user is specified, additional arguments can be supplied, in which case they are passed to the shell.

Which doesn't make it obvious how it behaves.

Personally I'd consider it a bug that two `--` might be required, but whatever...

Thanks for cross referencing it with upstream and util-linux. I should have done the same too first.

**Comment by ale5000-git** [comment]
Why not just stop options parsing at first non option arg or --?

**Comment by rmyorston** [comment]
>root before -c just seems wrong.

How it _seems_ is immaterial, it's _accepted_ by util-linux `su` and upstream BusyBox.

**Comment by ale5000-git** [comment]
With this:
`su -c 'echo "$0"; echo "$*"' -- root -- script_name -p param_options1 -q param_options2`
Are the parameters passed in the same order?
Otherwise everything break.

**Comment by avih** [comment]
> `root` before `-c` just seems wrong.

Not at all. The gist is that after the user (`root`), all arguments go directly to `sh`. It's not FILE [ARG...] like the busybox synopsis suggests, but rather what the util-linux manpage says - simply arguments to `sh`, and this is the most general invocation of `sh` in this context.

Using `-c` as an argument of `su`, rather than of `sh`, is redundant, and its only usefulness is to run a shell command without having to input the user name, and/or without using -- to stop the arguments parsing of `su`.
Generic where `-c` goes to `sh`: `su -- root -c CMD_STRING`
Identical but simpler to write: `su -c CMD_STRING`

So if `su` didn't support `-c`, then only the first form would work, you'd need both the user and the `--`.

**Comment by rmyorston** [comment]
The usage message has been fixed, the patch has been applied to master and a [prerelease](https://frippery.org/files/busybox/prerelease) is available.

Thanks for your patience while this got beaten into shape!

**Comment by avih** [comment]
> The usage message has been fixed

The behavior is good, I think, and that's currently the message as far as I can tell:
```
"[-c 'CMD'] [root [FILE ARGS | ARG0 ARGS]]"
```
But I think it's still misleading here and in upstream busybox, for two reasons:
1. It's not obvious that ARG0 ARGS is only relevant with -c CMD, and that FILE ARGS is only relevant without -c CMD.
2. It's not communicated at all that after `root` any additional shell arguments can be added before the ARG0/FILE part, and I think it's actually really important, because that's how it becomes a fully generic sh elevation invocation, rather than two specific modes: -c CMD... and FILE... .

The fact that it does behave like `su` in util-linux is great, but the busybox docs wrongly paint a different picture IMHO.

**Comment by avih** [comment]
Also, for reference, another enhancement I had (on top of my patches, so probably doesn't apply over current master) is to support `-w` for "wait for the process to exit and get its exit code". Here's the diff:
```diff
diff --git a/loginutils/suw32.c b/loginutils/suw32.c
index 14fda73c3..3b38287db 100644
--- a/loginutils/suw32.c
+++ b/loginutils/suw32.c
@@ -71,12 +71,13 @@ int suw32_main(int argc, char **argv) MAIN_EXTERNALLY_VISIBLE;
 int suw32_main(int argc UNUSED_PARAM, char **argv)
 {
 	SHELLEXECUTEINFO info;
-	unsigned opts, c_opt;
+	unsigned opts, c_opt, w_opt;
 	char *bb_path, *cwd;
 	DECLARE_PROC_ADDR(BOOL, ShellExecuteExA, SHELLEXECUTEINFOA *);
 
-	opts = getopt32(argv, "c");
+	opts = getopt32(argv, "cw");
 	c_opt = opts & 1;
+	w_opt = opts & 2;
 	if (!c_opt != !argv[optind])
 		bb_show_usage();  // -c witout CMD, or operand without -c
 
@@ -117,8 +118,23 @@ int suw32_main(int argc UNUSED_PARAM, char **argv)
 	/* info.lpDirectory = NULL; */
 	info.nShow = SW_SHOWNORMAL;
 
+	if (w_opt)
+		info.fMask |= SEE_MASK_NOCLOSEPROCESS;
+
 	if (!INIT_PROC_ADDR(shell32.dll, ShellExecuteExA))
 		return -1;
 
-	return !ShellExecuteExA(&info);
+	if (!ShellExecuteExA(&info))
+		return 1;
+
+	if (w_opt) {
+		WORD r;
+		WaitForSingleObject(info.hProcess, INFINITE);
+		if (!GetExitCodeProcess(info.hProcess, &r))
+			r = 1;
+		CloseHandle(info.hProcess);
+		return r;
+	}
+
+	return 0;
 }
```

**Comment by ale5000-git** [comment]
@rmyorston
It works properly, thank you :)

If possible it would be nice to also support the second `--`.
Two examples:
```
su -c 'echo "$0"; echo "$*"' -- root -- script_name
su -c 'echo "$0"; echo "$*"' -- -- script_name
```

**Comment by avih** [comment]
> If possible it would be nice to also support the second `--`.

`--` is not supported explicitly, but rather it's handled by the options parser of the respective utility, and indicates end-of-options (and begining of operands, if any).

Therefore a first `--` will be handled by the `su` options parser, and indicates that none of the following arguments are options to `su`.

In normal options parsing (POSIX guidelines, and POSIX `getopt`), `--` is not needed if `root` is given, because `root` doesn't begin with `-` and therefore interpreted as the first operand.

However, by default, util-linux, upstream busybox, and busybox-w32 permute the arguments of `su`, and "search" options even after the first argument which doesn't begin with `-`. And so, `--` right before or right after `root` will be handled by `su`.

> `su -c 'echo "$0"; echo "$*"' -- -- script_name`

You can't do `-- --` because after the `su` options, if there are any operands, the first must be the user, or, in busybox-w32, `root`. This is the same in util-linux `su`, and also in upstream busybox upstream `su`.

So the first `--` indicates end of `su` options, and, if `root` was not given, then the second `--` would be treated as the user, which is not `root`, hence rejected.

> `su -c 'echo "$0"; echo "$*"' -- root -- script_name`

This wouldn't work, because `-c` of `su` already adds the command string to the arguments it passed to `sh`, and `sh` treats it as the first operand, and so the second `--` is not needed and won't be interpreted by `sh` as "end of options" if prior operands were already supplied. When you use `-c` of `su`, what follows after `root` is `[script-name [arg...]]`.

So you simply don't need the second `--` here.

If you want generic invocation arguments to `sh`, then append them to `su -- root`, like you would for `sh`:
```sh
sh         -c -- 'echo "$0"; echo "$*"' script_name
su -- root -c -- 'echo "$0"; echo "$*"' script_name
```

**Comment by rmyorston** [comment]
I've taken some of the argument quoting code from this PR.

What would people like to see in the usage message?

The "wait for the process to exit and get its exit code" looks reasonable. But I see util-linux `su` already has a `w` option. What would be a good alternative?

**Comment by avih** [comment]
> I've taken some of the argument quoting code from this PR.

Sure. In mpv it uses quotes if the string has space/tab/double-quote. Apparently newline doesn't need quoting, but double-quote does trigger quoting. Not sure whether DQ value actually requires quoting.

> What would people like to see in the usage message?

Good question. Upstream has this, which I don't think is great either:
```
Usage: su [-lmp] [-s SH] [-] [USER [FILE ARGS | -c 'CMD' [ARG0 ARGS]]]
```

Maybe something along these lines?
```
Usage: su [root]
       su -c CMD_STRING [-- root [ARG0 [ARG...]]
       su -- root [arbitrary sh arguments]
```
It's a bit overlapping, but it shows that it can be used without arguments, or with CMD_STRING and then if args are needed then `-- root` is the bulletproof method, or with a fixed prefix followed by arbitrary arguments to sh.

> The "wait for the process to exit and get its exit code" looks reasonable. But I see util-linux `su` already has a `w` option. What would be a good alternative?

Hmm, I checked only upstream busybox `su`. Maybe `-e` for "exit code"?

**Comment by rmyorston** [comment]
>if the shell is already root, then it could execute the code without spawn another window

That's not unreasonable. Though this might be:
```diff
diff --git a/loginutils/suw32.c b/loginutils/suw32.c
index 34cb175ab..87b5616c9 100644
--- a/loginutils/suw32.c
+++ b/loginutils/suw32.c
@@ -41,6 +41,15 @@ int suw32_main(int argc UNUSED_PARAM, char **argv)
        ++argv;
    }
 
+   if (getuid() == 0) {
+       if (opt_command) {
+           *--argv = opt_command;
+           *--argv = (char *)"-c";
+       }
+       *--argv = (char *)"ash";
+       return ash_main(argc, argv);
+   }
+
    /* ShellExecuteEx() needs backslash as separator in UNC paths. */
    bb_path = xstrdup(bb_busybox_exec_path);
    slash_to_bs(bb_path);
```

**Comment by avih** [comment]
> That's not unreasonable. Though this might be:

Might be unreasonable? :)

What happens with `su -cecho` ? (didn't try too hard to follow the logic of this `*--argv=...` stuff).

**Comment by rmyorston** [comment]
>Might be unreasonable? :)

Yes, that was the implication. ;-)

>What happens with su -cecho ?

Right, we'd be pushing back two arguments when only one was consumed. And this:
```
    *--argv = xasprintf("-c%s", opt_command);
```
won't work because the shell doesn't like `-cecho`.  Hmm.

**Comment by avih** [comment]
Well, I don't think it should change the behavior, other than not-showing the UAC dialog if we're already elevated - which is the case already anyway.

Otherwise a script which uses `su` doesn't know if the command runs "in the background" or the foreground, which could definitely break the script IMO.

**Comment by ale5000-git** [comment]
A paramter like -n (like: no spawn if possible)?

**Comment by avih** [comment]
> A parameter like -n (like: no spawn if possible)?

What is the value in that?

The way I see it, there could be two cases:
1. You know you're already elevated, at which case why do you use `su` at all?
2. You don't know whether you're elevated, at which case why do you want the behavior to be unexpected?

**Comment by ale5000-git** [comment]
It is easier to test `su` if it doesn't spawn a new window every time.

**Comment by avih** [comment]
> It is easier to test `su` if it doesn't spawn a new window every time.

What does that mean? testing of what? If you're still trying to figure out how to pass arguments to `su`, then I already told you: just like on linux and with every other version of su, whatever arguments you want to use with an elevated `sh`, add the exact same arguments  after `su -- root`. That includes any `--` which you would have used with `sh`.

Also, it won't be `su` anymore, because it would simply run a new `sh`, and it will modify the arguments in some way, just like the suggested patch above tried to do, and ended up buggy.

**Comment by ale5000-git** [comment]
Is there a better way to test?
For example if I use `su root wrongtext` a new window appears and closes and I can't see anything.
This is just an example but maybe something to avoid closing the new window would be nice.

**Comment by avih** [comment]
> Is there a better way to test parameters?

I don't understand the question. 

Can you please describe a real world example where you would use such `-n` option?

**Comment by ale5000-git** [comment]
How I have said is for my personal testing, instead of a previous question what about a `-n` as not close the spawned window no error?

**Comment by avih** [comment]
> > The "wait for the process to exit and get its exit code" looks reasonable. But I see util-linux `su` already has a `w` option. What would be a good alternative?
> 
> Hmm, I checked only upstream busybox `su`. Maybe `-e` for "exit code"?

Maybe `-W` (capital)?

> How I have said is for my personal testing [...]

Sorry, I don't understand what you're saying, and you still didn't give any example of an actual scenario where `-n` would be used. Even if only for personal testing, can you give an example of how you would use that, from start to finish?

**Comment by rmyorston** [comment]
The [latest prerelease](https://frippery.org/files/busybox/prerelease) has two new flags:
- `-W` causes `su` to wait for the shell exit code. It's pretty much verbatim from the patch above.
- `-N` prevents the elevated console from closing when the shell exits. The shell issues a prompt and waits for user input.  Thus `su -N -- root wrongtext` displays:
   ```
   ash: can't open 'wrongtext': No such file or directory
   Press any key to exit...
   ```

**Comment by avih** [comment]
> The [latest prerelease](https://frippery.org/files/busybox/prerelease) has two new flags: ...

LGTM, however, we already know that here and in `su` elsewhere the `-c` option is redundant because it can also be passed to `sh` directly as part of the arbitrary arguments, but we do keep it because other `su` support it too.

But for `-N`, which is similarly redundant at `su`, why add more `su` code instead of only at `ash`?

Also, I'd think that the "press any key" message should be at stderr, in case something like `sudo` above tries to capture it.

(and I'll update `sudo` above to use `-W` instead of the poll loop)

**Comment by rmyorston** [comment]
>why add more su code instead of only at ash?

`ash` has a `-N` option? Who told you that? ;-)

>Also, I'd think that the "press any key" message should be at stderr, in case something like sudo above tries to capture it.

The message is written directly to the console (and the keystroke is read directly from the console) to work around pesky scripts redirecting things.

**Comment by avih** [comment]
Hmm.. I just recalled that `-W` can be special. Not sure how meaningful is that here...

> 12.2 Utility Syntax Guidelines
> ...
> Guideline 3:
>    Each option name should be a single alphanumeric character (the alnum character classification) from the portable character set. The -W (capital-W) option shall be reserved for vendor options.

**Comment by ale5000-git** [comment]
@rmyorston
It is working perfectly, thanks.

@avih
> Sorry, I don't understand what you're saying, and you still didn't give any example of an actual scenario where `-n` would be used. Even if only for personal testing, can you give an example of how you would use that, from start to finish?

Sorry for the late reply.
I don't have a real example because I do many things at the same time and I don't yet had time to do a  things that need su.
But often I write scripts to automate things on Windows and if there is a typo in my code it is very useful to see the error message to speed up the fix; a single line of code could be very long so the error message really helps.
The -N option already added is perfect.

**Comment by ale5000-git** [comment]
Is there anything in this PR that isn't already in master?

**Comment by rmyorston** [comment]
I don't think so.

In a week or two I'll review all open issues (and PRs) that should have been fixed in the FRP-5007 release. If there haven't been any further comments I'll close them.

**Comment by avih** [comment]
I'm happy with current master.

**Comment by avih** [comment]
becedd5bd29dd1ea3c21426568e6412d6b9a404e looks good.

What do you think about the possible xargs issue?

**Comment by rmyorston** [comment]
>What do you think about the possible xargs issue?

Hope to release something shortly, it's currently in testing.

**Comment by rmyorston** [comment]
Try the [latest prerelease](https://frippery.org/files/busybox/prerelease) (PRE-5535 or above).

I've been testing with this:
```
dd if=/dev/zero bs=1000 count=100 | xargs -0 uptime
```
The command run by `xargs` has to be spawned for there to be a problem. If it's a NOFORK applet it's invoked directly, so `echo`, for example, doesn't display the problem.

**Comment by avih** [comment]
> Try ...

Not soon, but I trust your tests. It's been a while since I've updated my build, and I'm postponing it because I need to rebase my patches... but I really should get to that...

Re commit dc3ee8e0523a4f651f3012ae1f5a181548187709 :

So did it fix a real issue which you were able to reproduce? I.e. was that indeed an xargs bug on windows?

> found that `_spawnv` works with at least 32759 chars combined args length on XP32, win7-64, and win10-64 (1290 args)

> `# define WIN32_MAX_CHARS (32750)`

I might have been unclear. 32759 is only the number which I confirmed works (which was 1290 arguments which my loop generated, and it didn't work with 1291 args which were total more than 32768).

I presume the exact limit is 32767 or some such, although it might also depend on which exact API is used, and how wrong the docs are...

> `char *s = quote_arg(arg);`

While probably not a critical performance issue (especially if only applied to the original args, and not to the concatenated list which xargs can generate, which I presume is the case), this still feels wasteful to me, because it runs `xmalloc` and maybe `memmove`, while it can be implemented using a small and simple one-pass loop which only counts.

Also, without trying to study `xargs_main`, that "crop list, save tail for later" thingy feels weird to me.

I presume it was implemented like that to avoid big windows diff inside the actual xargs logic? because otherwise, I feel that there must be a nicer solution instead of this...

**Comment by rmyorston** [comment]
@avih Yes, there is a real issue. Nobody has complained about it, probably because the 2,048 character reduction POSIX makes to the maximum length allows for a reasonable amount of expansion on Windows.

The performance of `xargs` in the latest prerelease is the same as before the recent changes. As usual, I prefer to optimise for size rather than speed. With your `quote_len()` and some other changes I can get a 1.5% increase in performance, but with an additional 20% of bloat.

**Comment by avih** [comment]
> additional 20% of bloat

20% of what?
